numpy/core/src/private/binop_override.h

```diff
@@ -0,0 +1,292 @@
+ *   all use cases. (See https://github.com/numpy/numpy/issues/5844 for lots
+ *   of discussion of this, and in particular
+ *     https://github.com/numpy/numpy/issues/5844#issuecomment-112014014
+ *   for a summary of some conclusions.)
+ *
+ * So for 1.10, we are going to try the following rules. a.__add__(b) will
+ * be implemented as follows:
+ * - If b does not define __numpy_ufunc__, apply the legacy rule:
+ *   - If not isinstance(b, a.__class__), and b.__array_priority__ is higher
+ *     than a.__array_priority__, return NotImplemented
+ *   - Otherwise, fall through.
+ * - If b->ob_type["__module__"].startswith("scipy.sparse."), then return
+ *   NotImplemented. (Rationale: scipy.sparse defines __mul__ and np.multiply
+ *   to do two totally different things. We want to grandfather this behavior
+ *   in, but we don't want to support it in the long run, as per PEP
+ *   465. Additionally, several versions of scipy.sparse were released with
+ *   __numpy_ufunc__ implementations that don't match the final interface, and
+ *   we don't want dense + sparse to suddenly start erroring out because
+ *   dense.__add__ dispatched to a broken sparse.__numpy_ufunc__.)
+ * - Otherwise, call the corresponding ufunc.
+ *
+ * For reversed operations like b.__radd__(a), and for in-place operations
+ * like a.__iadd__(b), we:
+ * - Call the corresponding ufunc
+ *
+ * Rationale for __radd__: This is because by the time the reversed operation
+ * is called, there are only two possibilities: The first possibility is that
+ * the current class is a strict subclass of the other class. In practice, the
+ * only way this will happen is if b is a strict subclass of a, and a is
+ * ndarray or a subclass of ndarray, and neither a nor b has actually
+ * overridden this method. In this case, Python will never call a.__add__
+ * (because it's identical to b.__radd__), so we have no-one to defer to;
+ * there's no reason to return NotImplemented. The second possibility is that
+ * a.__add__ has already been called and returned NotImplemented. Again, in
+ * this case there is no point in returning NotImplemented.
+ *
+ * Rationale for __iadd__: In-place operations do not take all the trouble
+ * above, because if __iadd__ returns NotImplemented then Python will silently
+ * convert the operation into an out-of-place operation, i.e. 'a += b' will
+ * silently become 'a = a + b'. We don't want to allow this for arrays,
+ * because it will create unexpected memory allocations, break views,
+ * etc.
+ *
+ * In the future we might change these rules further. For example, we plan to
+ * eventually deprecate __array_priority__ in cases where __numpy_ufunc__ is
+ * not present, and we might decide that we need somewhat more flexible
+ * dispatch rules where the ndarray binops sometimes return NotImplemented
+ * rather than always dispatching to ufuncs.
+ *
+ * Note that these rules are also implemented by ABCArray, so any changes here
```

Comment:
Hmm... no commitments, but given that I've already signed up to write `BackwardCompatibleABCArray` (outside of numpy) I suppose I can give this a try.